recipe_scrapers/_schemaorg.py

```diff
@@ -2,18 +2,58 @@
 # IF things in this file continue get messy (I'd say 300+ lines) it may be time to
 # find a package that parses https://schema.org/Recipe properly (or create one ourselves).
 
+from collections import defaultdict
+import json
 
-import extruct
-
-from recipe_scrapers.settings import settings
+import html5lib
 
 from ._exceptions import SchemaOrgException
 from ._utils import get_minutes, get_yields, normalize_string
 
-SCHEMA_ORG_HOST = "schema.org"
-SCHEMA_NAMES = ["Recipe", "WebPage"]
 
-SYNTAXES = ["json-ld", "microdata"]
+class Extractor:
+    def __init__(self, page_data):
+        self.tree = html5lib.parse(page_data)
+
+    @property
+    def linked_data(self):
+        for element in self.tree.findall(
+            path=".//script[@type='application/ld+json']",
+            namespaces={"": "http://www.w3.org/1999/xhtml"},
+        ):
+            try:
+                yield json.loads(element.text, strict=False)
+            except Exception:
+                pass
+
+    @property
+    def microdata(self):
+        yield from self.tree.findall(
+            path=".//*[@itemscope]",
+            namespaces={"": "http://www.w3.org/1999/xhtml"},
+        )
+
+
+def chunked_text(node):
+    if node.text:
+        yield node.text
+    if isinstance(node.tag, str) and node.tag.endswith("}br"):
+        yield "\n"
+    for child in node:
+        prev = None
+        for chunk in chunked_text(child):
+            if chunk == prev == "\n":
+                continue
+            yield chunk
+            prev = chunk
+    if node.tail:
+        if node.tail[0].isspace():
+            trimmed = node.tail.lstrip()
+            if trimmed and not trimmed[0] == ",":
```

Comment:
This should really be a "character is punctuation" check.

Roughly speaking, whitespace removal for `first <div>foo</div> second` vs `first <div>foo</div>   , second` is expected to occur differently according to our test cases.